nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,490 @@
+	var url string
+	timeout := d.Get("bundle_upload_timeout").(int)
+	c := m.(nsxtClients)
+	userName := c.NsxtClientConfig.UserName
+	password := c.NsxtClientConfig.Password
+	connector := getPolicyConnector(m)
+	summaryClient := upgrade.NewSummaryClient(connector)
+	summary, err := summaryClient.Get()
+	if err != nil {
+		return err
+	}
+	if bundleType == nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_UPGRADE {
+		if !d.HasChange("upgrade_bundle_url") {
+			return nil
+		}
+		url = upgradeBundleURL
+	} else {
+		// Check if same precheck bundle has already been uploaded by comparing input target version
+		// with the precheck bundler version in upgrade summary
+		// version format is x.x.x...., contains at least 3 numbers
+		version := d.Get("version").(string)
+		if summary.PreUpgradeBundleVersion != nil && version != "" {
+			existingBundleVersion := *summary.PreUpgradeBundleVersion
+			arr := strings.Split(version, ".")
+			if len(arr) < 3 {
+				log.Printf("Invalid input version format, cannot compare version, input version format must be x.x.x....")
+			} else {
+				if strings.HasPrefix(existingBundleVersion, version) {
+					log.Printf("Precheck bundle already uploaded to NSX, skip uploading precheck bundle")
+					return nil
+				}
+			}
+		}
+		url = precheckBundleURL
+	}
+	client := upgrade.NewBundlesClient(connector)
+	bundleFetchRequest := nsxModel.UpgradeBundleFetchRequest{
+		Url: &url,
+	}
+	if nsxVersionHigherOrEqual("4.1.1") {
+		bundleFetchRequest.BundleType = &bundleType
+		bundleFetchRequest.Password = &password
+		bundleFetchRequest.Username = &userName
+	}
+	bundleID, err := client.Create(bundleFetchRequest, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to upload upgrade bundle of type %s: %v", bundleType, err)
+	}
+	return waitForBundleUpload(m, *bundleID.BundleId, timeout)
+}
+
+func acceptUserAgreement(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	acceptUserAgreement := d.Get("accept_user_agreement").(bool)
+	if !acceptUserAgreement {
+		return fmt.Errorf("To proceed with upgrade, you must accept user agreement")
+	}
+	client := eula.NewAcceptClient(connector)
+	err := client.Create()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeUc(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	summaryClient := upgrade.NewSummaryClient(connector)
+	summary, err := summaryClient.Get()
+	if err != nil {
+		return err
+	}
+	if *summary.UpgradeCoordinatorUpdated {
+		log.Printf("Upgrade coordinator already upgraded")
+		return nil
+	}
+	client := nsx.NewUpgradeClient(connector)
+	err = client.Upgradeuc()
+	if err != nil {
+		return err
+	}
+	timeout := d.Get("uc_upgrade_timeout").(int)
+	return waitForUcUpgrade(m, timeout)
+}
+
+func executePreupgradeChecks(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
```

Comment:
Lets add some logging here, and also in `precheckNeeded` routine so that we know why precheck was triggered or not triggered.

Reply:
done